istioctl/pkg/kubernetes/client.go

```diff
@@ -0,0 +1,200 @@
+type Client struct {
+	Config *rest.Config
+	*rest.RESTClient
+}
+
+// NewClient is the contructor for the client wrapper
+func NewClient(kubeconfig, configContext string) (*Client, error) {
+	config, err := defaultRestConfig(kubeconfig, configContext)
+	if err != nil {
+		return nil, err
+	}
+	restClient, err := rest.RESTClientFor(config)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{config, restClient}, nil
+}
+
+func defaultRestConfig(kubeconfig, configContext string) (*rest.Config, error) {
+	config, err := kube.BuildClientConfig(kubeconfig, configContext)
+	if err != nil {
+		return nil, err
+	}
+	config.APIPath = "/api"
+	config.GroupVersion = &v1.SchemeGroupVersion
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
+	return config, nil
+}
+
+// PodExec takes a command and the pod data to run the command in the specified pod
+func (client *Client) PodExec(podName, podNamespace, container string, command []string) (*bytes.Buffer, *bytes.Buffer, error) {
+	req := client.Post().
+		Resource("pods").
+		Name(podName).
+		Namespace(podNamespace).
+		SubResource("exec").
+		Param("container", container).
+		VersionedParams(&v1.PodExecOptions{
+			Container: container,
+			Command:   command,
+			Stdin:     false,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(client.Config, "POST", req.URL())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var stdout, stderr bytes.Buffer
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+		Tty:    false,
+	})
+
+	return &stdout, &stderr, err
+}
+
+// AllPilotsDiscoveryDo makes an http request to each Pilot discovery instance
+func (client *Client) AllPilotsDiscoveryDo(pilotNamespace, method, path string, body []byte) (map[string][]byte, error) {
+	pilots, err := client.GetPilotPods(pilotNamespace)
+	if err != nil {
+		return nil, err
+	}
+	if len(pilots) == 0 {
+		return nil, errors.New("unable to find any Pilot instances")
+	}
+	cmd := []string{"/usr/local/bin/pilot-discovery", "request", method, path, string(body)}
+	var stdout, stderr *bytes.Buffer
+	result := map[string][]byte{}
+	for _, pilot := range pilots {
+		stdout, stderr, err = client.PodExec(pilot.Name, pilot.Namespace, discoveryContainer, cmd)
+		if err != nil {
+			return nil, fmt.Errorf("error execing into %v %v container: %v", pilot.Name, discoveryContainer, err)
+		} else if stdout.String() != "" {
+			result[pilot.Name] = stdout.Bytes()
+		} else if stderr.String() != "" {
+			return nil, fmt.Errorf("error execing into %v %v container: %v", pilot.Name, discoveryContainer, stderr.String())
+		}
+	}
+	return result, err
+}
+
+// PilotDiscoveryDo makes an http request to a single Pilot discovery instance
+func (client *Client) PilotDiscoveryDo(pilotNamespace, method, path string, body []byte) ([]byte, error) {
```

Comment:
Do we need this one (or have separate AllPilotDo and PilotDo). I guess it would be no harm to use AllPilotDiscoveryDo all the time, and it will return the same result if there is only one pilot?

Reply:
The aim was to keep as much of the logic out of `istioctl/cmd/*` as possible so that things are easier to test. I realise that this file isn't currently tested but I plan on circling back around to write some tests for this at some point.